src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -41,10 +42,19 @@ public SocketInput(MemoryPool memory, IThreadPool threadPool, IBufferSizeControl
             _awaitableState = _awaitableIsNotCompleted;
         }
 
-        public bool RemoteIntakeFin { get; set; }
-
         public bool IsCompleted => ReferenceEquals(_awaitableState, _awaitableIsCompleted);
 
+        internal bool ReadingInput => _tcs.Task.Status == TaskStatus.WaitingForActivation;
+
+        public bool CheckFinOrThrow()
+        {
+            lock (_sync)
```

Comment:
I think this was only required when we stored `_awaitableError` and `RemoteIntakeFin` in separate fields.

We can't just remove the lock with the code as-is, because there is the risk that this would return true after the connection is faulted. I think the lock could be removed if the code changed to:

    CheckConnectionError();
    return _tcs.Task.Status == TaskStatus.RanToCompletion;